llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,785 @@
+
+  while (CurrentPtr < Tokens.size()) {
+    Token CurrentToken = Tokens[CurrentPtr];
+    CurrentPtr++;
+    Accessor A = CurrentToken.getAccessor();
+    ASTNode *CurrentNode;
+    void *Node = Allocator.Allocate(sizeof(ASTNode), alignof(ASTNode));
+
+    switch (CurrentToken.getType()) {
+    case Token::Type::Text: {
+      CurrentNode =
+          CreateTextNode(Node, CurrentToken.getTokenBody(), Parent, Alloc,
+                         Partials, Lambdas, SectionLambdas, Escapes);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::Variable: {
+      CurrentNode = CreateNode(Node, ASTNode::Variable, A, Parent, Alloc,
+                               Partials, Lambdas, SectionLambdas, Escapes);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::UnescapeVariable: {
+      CurrentNode =
+          CreateNode(Node, ASTNode::UnescapeVariable, A, Parent, Alloc,
+                     Partials, Lambdas, SectionLambdas, Escapes);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::Partial: {
+      CurrentNode = CreateNode(Node, ASTNode::Partial, A, Parent, Alloc,
+                               Partials, Lambdas, SectionLambdas, Escapes);
+      CurrentNode->setIndentation(CurrentToken.getIndentation());
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::SectionOpen: {
+      CurrentNode = CreateNode(Node, ASTNode::Section, A, Parent, Alloc,
+                               Partials, Lambdas, SectionLambdas, Escapes);
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode, Alloc, Partials, Lambdas, SectionLambdas,
+                    Escapes);
+      const size_t End = CurrentPtr - 1;
+      std::string RawBody;
+      for (std::size_t I = Start; I < End; I++)
+        RawBody += Tokens[I].getRawBody();
+      CurrentNode->setRawBody(std::move(RawBody));
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::InvertSectionOpen: {
+      CurrentNode = CreateNode(Node, ASTNode::InvertSection, A, Parent, Alloc,
+                               Partials, Lambdas, SectionLambdas, Escapes);
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode, Alloc, Partials, Lambdas, SectionLambdas,
+                    Escapes);
+      const size_t End = CurrentPtr - 1;
+      std::string RawBody;
+      for (size_t Idx = Start; Idx < End; Idx++)
+        RawBody += Tokens[Idx].getRawBody();
+      CurrentNode->setRawBody(std::move(RawBody));
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::Comment:
+      break;
+    case Token::Type::SectionClose:
+      return;
+    }
+  }
+}
+void toMustacheString(const Value &Data, raw_ostream &OS) {
+  switch (Data.kind()) {
+  case Value::Null:
+    return;
+  case Value::Number: {
+    auto Num = *Data.getAsNumber();
+    std::ostringstream Oss;
+    Oss << Num;
```

Comment:
nit: lets rename to `SS`. I just misread this code as using the same variable, so lets try to avoid making that kind of mistake too easy.